Optimizations, plus some cleanups and enhancements.

jsiembida opened the pull request:
Optimizations:

- Use Float32 typed arrays for physics and rendering
- Use multi threaded worker pool for rendering
- Remove subsampling in rendering code, it is fast enough for full resolution
- Firefox now seems to perform slightly better than Chrome (code was not optimized for any specific browser though)
- On desktop with 8 rendering threads, FF and Chrome do around 60fps for 20 charges.
  At this point the physics part becomes a bottleneck, especially for higher supersamplings.
  On MBP 2015, FF and Chrome do 10-15fps for 20 charges, while Safari does ~20fps.
- FF on macOS incorrectly reports number of physical instead of logical CPU cores
  so it performes worst of the three in this case.

Cleanups:

- Remove global states, pass everything in messages
- Fix HTML input handling
- Fix canvas size handling
- Remove firefox specific stuff, no longer needed

Enhancements:

- Add mass / inertia of the charge, use this to provide the static charge option
- Add "pause" option, rename "timescale" to "speed"
- Add 800x800 resolution
- More on-screen stats

Commit:
Optimizations, plus some cleanups and enhancements.

Optimizations:

- Use Float32 typed arrays for physics and rendering
- Use multi threaded worker pool for rendering
- Remove subsampling in rendering code, it is fast enough for full resolution
- Firefox now seems to perform slightly better than Chrome (code was not optimized for any specific browser though)
- On desktop with 8 rendering threads, FF and Chrome do around 60fps for 20 charges.
  At this point the physics part becomes a bottleneck, especially for higher supersamplings.
  On MBP 2015, FF and Chrome do 10-15fps for 20 charges, while Safari does ~20fps.
- FF on macOS incorrectly reports number of physical instead of logical CPU cores
  so it performes worst of the three in this case.

Cleanups:

- Remove global states, pass everything in messages
- Fix HTML input handling
- Fix canvas size handling
- Remove firefox specific stuff, no longer needed

Enhancements:

- Add mass / inertia of the charge, use this to provide the static charge option
- Add "pause" option, rename "timescale" to "speed"
- Add 800x800 resolution
- More on-screen stats

Commit:
Use plain loops everywhere in physics and render code.

Chrome really hated the reduce, it seemed ok for FF.
Now both do well, and Chrome is faster now at rendering.
Both can sustain 60fps and 30 particles at 600x600.

Commit:
Fix jitter in fps stat, add render pool load stat, fix race condition when adding a charge.

Commit:
Micro optimization, manually inlined inner loop in physics.

Commit:
More micro optimizations for physics.

Commit:
Less agressive threading, slower simulation speed.

Commit:
Use normalized coordinates for simulation. -1.0 ... 1.0 for x and y. Adjust scaling factors accordingly.

This makes simulation elements insensitive to canvas size.
The particles are visualized in relation to canvas size.
Resizing canvas during simulation is possible now as long as ratio is preserved.
Potentially, WebGL could be used to render the scene later on (WebGL clip space is also -1.0 ... 1.0).

Commit:
Adding initial WASM implementations for both physics and rendering.

Commit:
Change the handling of medium friciton factor in simulation.

Before: friction was only applied after subsampling steps were done in simulation.
After: friction is applied in every subsampling step.

While the change seems more correct it introduces rounding errors.
It still "feels less wrong" - proper error analysis might prove
the feeling completely wrong though.

JoshGreenslade commented:
These are some pretty stunning optimizations, way beyond anything I thought was possible! Awesome stuff

RE the rounding errors with the friction applied at every step, I reckon at this speed we can switch from doing Euler integration to doing some proper update steps using RK4: https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods

Additionally, we can add in the repulsive force when the particles get too close to prevent the weird physics that happens (though actual weird physics does happen in the real world when this happens, just not the kind we're observing here).

Rather than do a merge and have people think I did all this, I think I'm more likely to keep this on a separate branch and upload it as a separate update to the field on a different page. That way we can make sure you get credited for the work?

jsiembida commented:
Hello, I am ok with whatever you decide re the repo, at some point I will merge it to master in my fork once I complete a few more changes. As for credits, GH holds history, no problem here. Re the simulation details, a couple of observations:

1. IMHO, there is one more inaccuracy in the simulation code. New state of a particle is calculated based on partially updated state. That is, for particles numbered 0...M...N, particleM is calculated with particles 0...M-1 updated for time T+1 and particles M+1...N still for T. That doesn't look right irrespective of the integration method.

2. I was thinking about different integration methods, was looking first at the midpoint one as it is simple and still should improve things. With some refactor, code could be swappable underneath with a UI option to choose between them.

3. But I think major performance issue is in quadratic complexity of the formulas the simulation is based on.

Commit:
A bunch of changes to simulation parameters and UI controls.

- Naming more obvious, let's be blatantly obvious.
- Parametrized elasticity of walls.
- Simulation parameters values normalized, their actual values are kinda meaningless anyway,
  user needs to see the relative values.
- Slider for particle charges now goes to negative values.
- Better pause / play handling with a dedicated button.
- Reset button resets everything.
- Dedicated class to handle parameters. (will come handy in implementing simulation save / load later on)
- Controls put into a table so they get nicely aligned.

Commit:
Change particle state handling in physics code path.

Create two working buffer sets, current and next.
Calculate new particle states into the next buffer set using current set.
Flip the buffers at the end of iteration.

Commit:
More integration methods and load / save functionalities.

- Added Midpoint, Heun and Runge-Kutta (currently only implementation in Javascript).
- Integration method choice is exposed in UI.
- Added load / save buttons so that simulation state can be persisted in a json file.

Commit:
Add "single step" button.

Commit:
Adaptive time scaling, layout grid and some fixes.

- Adaptive time scaling changes dt depending on max velocity of particles.
- Layout grid to "snap" new particles as they are being added, handy when creating symmetric layouts.
- Fixed the loading of a saved simulation.

Commit:
Rework the layout of the page, hopefully more responsive, only 3 break points, landscape and portrait, etc.

Commit:
Rework stats interface. Fix WASM/MIME issue. Styling fixes.

Commit:
Verlet integration method.

Commit:
Switching to actual physical units, added indication of system kinetic and potential energy. And another rework of configuration UI.

Numbers run in physics code now represent:
- Simulation scene positions are in meters. From -1m to 1m in both dimensions.
- Charges are in coulombs.
- Time is in seconds.
- Mass in kilograms.
- In result we get velocities in meters / sec and energies in joules.
